posthog/temporal/batch_exports/transformer.py

```diff
@@ -55,243 +51,345 @@ def get_stream_transformer(
+                future = loop.run_in_executor(
+                    executor, dump_record_batch, record_batch, self.compression, self.include_inserted_at
+                )
+                self._futures_pending.add(future)
 
-    @property
-    def brotli_compressor(self):
-        if self._brotli_compressor is None:
-            self._brotli_compressor = brotli.Compressor()
-        return self._brotli_compressor
+        producer_task = asyncio.create_task(producer())
 
-    def finish_brotli_compressor(self) -> typing.Generator[bytes, None, None]:
-        """Flush remaining brotli bytes."""
-        if self.compression != "brotli":
-            raise ValueError(f"Compression is '{self.compression}', not 'brotli'")
+        try:
+            yield producer_task
+        finally:
+            if not producer_task.done():
+                _ = producer_task.cancel()
 
-        yield self.brotli_compressor.finish()
-        self._brotli_compressor = None
+            try:
+                await producer_task
+            except asyncio.CancelledError:
+                pass
 
 
-class JSONLStreamTransformer(StreamTransformer):
-    def _write_dict(self, d: dict[str, typing.Any]) -> typing.Generator[bytes, None, None]:
-        """Write a single row of JSONL."""
-        try:
-            data_gen = self._write(orjson.dumps(d, default=str) + b"\n")
-        except orjson.JSONEncodeError as err:
-            # NOTE: `orjson.JSONEncodeError` is actually just an alias for `TypeError`.
-            # This handler will catch everything coming from orjson, so we have to
-            # awkwardly check error messages.
-            if str(err) == "Recursion limit reached":
-                # Orjson enforces an unmodifiable recursion limit (256), so we can't
-                # dump very nested dicts.
-                if d.get("event", None) == "$web_vitals":
-                    # These are PostHog events that for a while included a bunch of
-                    # nested DOM structures. Eventually, this was removed, but these
-                    # events could still be present in database.
-                    # Let's try to clear the key with nested elements first.
-                    try:
-                        del d["properties"]["$web_vitals_INP_event"]["attribution"]["interactionTargetElement"]
-                    except KeyError:
-                        # We tried, fallback to the slower but more permissive stdlib
-                        # json.
-                        logger.exception("PostHog $web_vitals event didn't match expected structure")
-                        dumped = json.dumps(d, default=str).encode("utf-8")
-                        data_gen = self._write(dumped + b"\n")
-                    else:
-                        dumped = orjson.dumps(d, default=str)
-                        data_gen = self._write(dumped + b"\n")
+class JSONLBrotliStreamTransformer:
```

Comment:
could we not subclass the `JSONLStreamTransformer`? I haven't looked closely but seems like most of the code is very similar

Reply:
I would rather err on the side of composition instead of inheriting. The difficulty in adapting to a common interface is what lead to the creation of a protocol in the first place.

That being said, I do see an opportunity to remove duplication in the `_record_batches_producer` method that both `JSON*` and `JSONL*` define, so I've moved that to a function.